chain/chain/src/chain.rs

```diff
@@ -5023,6 +5014,18 @@ impl<'a> ChainUpdate<'a> {
                     sum_gas_used += gas_burnt;
                     sum_balance_burnt += balance_burnt;
 
+                    if let Some(manager) = self.runtime_adapter.get_flat_storage_manager() {
+                        let store_update = manager.save_flat_state_changes(
+                            *block.hash(),
+                            *block.header().prev_hash(),
+                            block.header().height(),
+                            result.shard_uid,
+                            result.trie_changes.state_changes(),
+                        )?;
+                        manager.update_flat_storage_for_shard(*shard_uid, block)?;
```

Comment:
Sanity check: 
What should be the order of the following two events:
1) update_flat_storage_for_shard()
2) chain_store_update.commit() 

In postprocess_block it looks like we do 2) first from the postprocess_block_only and 1) second.
In block_catch_up_postprocess it looks like we do 2) first and 1) second. 

Here, if I'm not wrong, you do 1) first and 2) second.

Reply:
You're right, good catch! Seems like in all places we first commit the flat state changes and then call `update_flat_storage_for_shard`.

Looked up the code and seems like the consequence of current code (where we are committing later) would be that in updating head we wouldn't be seeing the latest delta as it's not committed.

Reply:
I'll try to improve documentation here. But would love for @Longarithm to confirm the change/issue here :)

Reply:
This problem is deep 😭😭😭😭😭 

Application of flat storage changes is very disconnected with updating flat storage head. It happens in two places, one during catchup, one during block processing.

The catchup one is easy to reason about. Mainly the `block_catch_up_postprocess` function where `chain_update.apply_chunk_postprocessing` writes the deltas and later we do a `manager.update_flat_storage_for_shard`.

The other is `postprocess_block` function where we apply all changes from a list of blocks in `self.postprocess_block_only` and then later update head. Any of the `apply_results` can have split state processing.

I feel like the fix here is to have saving flat state deltas coupled with updating flat storage head. I've create an issue for it here https://github.com/near/nearcore/issues/9430

Meanwhile for this change, since we are never reading from the child shards, while we fix the issue above, it might be fine to leave the code as is.

The consequence, as I mentioned, would be that we do not apply the changes from latest deltas. @Longarithm @pugachAG, thoughts?

Reply:
My preference here would be to merge PR as is. I've added a TODO with the issue details for the fix which is non trivial.

Reply:
I was hoping you can fix that by moving the call to update_flat_storage_for_shard higher up the call stack (or lower depending how you look). Check this out, ideally you can move it from here 

https://miro.com/app/board/uXjVMuVLtzo=/?share_link_id=781116851334

Reply:
Discussed offline, it's not that bad, it only makes the flat storage to be one block behind and everything should still work as expected. We will follow up later.